### PR TITLE
Add NotNextTo loss strategy and solver

### PR DESCRIPTION
**Pull request by xyao-nv**
## Summary
Add NotNextTo loss strategy and solver

## Detailed description
- Why:  when a task goal is `NextTo(parent, side, distance)` like Place A on +x side of B. The init state must NOT already satisfy it.

- What: Added  `NotNextTo` strategy, which is basically staying away from the entire whole side of the parent, and within parent's perpendicular footprint.
- The loss funcs's derivative is to either cross back over the parent's, or step step out past either side of parent's footprint.

image

### CODE REVIEW

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 1]
```diff
@@ -319,6 +319,184 @@ def compute_loss(
...
+
+        # Inside-band penetration: distance from child center to the
+        # nearer edge, clamped to >= 0. Zero when the child has
+        # escaped the band along that axis.
+        zero = torch.zeros((), dtype=child_pos.dtype, device=child_pos.device)
+        inside_x = torch.maximum(zero, torch.minimum(child_pos[:, 0] - valid_x_min, valid_x_max - child_pos[:, 0]))
+        inside_y = torch.maximum(zero, torch.minimum(child_pos[:, 1] - valid_y_min, valid_y_max - child_pos[:, 1]))
+
+        # min(): a single-axis escape is enough to satisfy Not(On).
+        loss = self.slope * torch.minimum(inside_x, inside_y)
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relation_loss_strategies.py]: Suggestion to move the all loss computation into `loss_primitives.py` into respective functions next to the others.

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 2]
```diff
@@ -319,6 +319,184 @@ def compute_loss(
...
+        escape_dist = torch.abs(primary - target_pos)
+
+        # Per-condition "how much of the margin is unfilled". Zero once that escape passes margin.
+        margin = self.margin_m
+        gap_side = torch.maximum(zero, margin - escape_side)
+        gap_cross = torch.maximum(zero, margin - escape_cross)
+        gap_dist = torch.maximum(zero, margin - escape_dist)
+
+        # min(): a single escape past the margin is enough to satisfy Not(NextTo).
+        loss = self.slope * torch.minimum(torch.minimum(gap_side, gap_cross), gap_dist)
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relation_loss_strategies.py]: Same here, suggestion to move the loss computation into `loss_primitives.py` to line this up stylistically with the existing strategies.

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 3]
```diff
@@ -319,6 +319,184 @@ def compute_loss(
...
+            child_offset = child_bbox.min_point[:, cfg.primary_axis]
+        else:
+            parent_edge = parent_world_bbox.min_point[:, cfg.primary_axis]
+            child_offset = child_bbox.max_point[:, cfg.primary_axis]
+        target_pos = parent_edge + cfg.direction * distance - child_offset
+
+        # Cross band: child placed at target position within parent's perpendicular extent.
+        parent_band_min = parent_world_bbox.min_point[:, cfg.band_axis]
+        parent_band_max = parent_world_bbox.max_point[:, cfg.band_axis]
+        valid_band_min = parent_band_min - child_bbox.min_point[:, cfg.band_axis]
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relation_loss_strategies.py:377]: I think we could reuse some existing losses like the `linear_band_loss()` from `loss_primitives.py` in this function

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 4]
```diff
@@ -319,6 +319,184 @@ def compute_loss(
...
+        valid_band_max = parent_band_max - child_bbox.max_point[:, cfg.band_axis]
+
+        primary = child_pos[:, cfg.primary_axis]
+        cross = child_pos[:, cfg.band_axis]
+        zero = torch.zeros((), dtype=child_pos.dtype, device=child_pos.device)
+
+        # escape_side: how far on the WRONG side of parent's edge (0 if on correct side).
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relation_loss_strategies.py]: This looks like the half-plane logic that `single_boundary_linear_loss()` implements

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 1 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relation_loss_strategies.py]: done

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 4 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relation_loss_strategies.py]: done

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 2 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relation_loss_strategies.py]: done

File: isaaclab_arena/relations/relation_loss_strategies.py (hunk 3 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relation_loss_strategies.py:377]: done

File: isaaclab_arena/relations/relation_loss_strategies.py [hunk 5]
```diff
@@ -319,6 +319,126 @@ def compute_loss(
...
+        #    extent (inverse of NextTo's band loss; 0 inside the band).
+        escape_cross = linear_band_loss(cross, valid_band_min, valid_band_max, slope=1.0)
+        gap_cross = single_boundary_linear_loss(escape_cross, self.margin_m, slope=1.0, penalty_side="less")
+
+        # 3. Distance escape: how far the primary-axis position is from the target distance
+        #    (inverse of NextTo's distance loss; 0 exactly at the target).
+        escape_dist = single_point_linear_loss(primary, target_pos, slope=1.0)
+        gap_dist = single_boundary_linear_loss(escape_dist, self.margin_m, slope=1.0, penalty_side="less")
+
+        # min(): a single escape past the margin is enough to satisfy Not(NextTo).
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relation_loss_strategies.py]: This doesn't exist `Not(NextTo)` no?

**Comment by xyao-nv** [comment]
Updated loss term as the cheaper of 2 escape distances, either going sideways off the footprint, or back over the edge.

File: isaaclab_arena/relations/relations.py [hunk 6]
```diff
@@ -142,6 +142,35 @@ def __init__(
 
 
 @register_object_relation
```
Comment (cvolkcvolk) [inline isaaclab_arena/relations/relations.py:144]: As greptile states below, the decorator belongs to `class IsAnchor(RelationBase):`

**Review by cvolkcvolk — APPROVED** [review]

File: isaaclab_arena/relations/relations.py (hunk 6 above)
Comment (xyao-nv) [reply isaaclab_arena/relations/relations.py:144]: added

File: isaaclab_arena/relations/relations.py [hunk 7]
```diff
@@ -142,6 +142,35 @@ def __init__(
...
+        Args:
+            parent: The parent asset whose adjacent half-plane is forbidden.
+            relation_loss_weight: Weight for the relationship loss function.
+            side: Which side of the parent is blocked (default: Side.POSITIVE_X).
+        """
+        super().__init__(parent, relation_loss_weight)
+        self.side = side
+
+
 class IsAnchor(RelationBase):
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/relations/relations.py:175]: added

File: isaaclab_arena/tests/test_relation_loss_strategies.py [hunk 8]
```diff
@@ -260,3 +263,179 @@ def test_next_to_loss_strategy_multi_env_shape_and_values():
...
+    child_pos = torch.tensor([[1.05, 0.4, 0.5], [0.3, 0.4, 0.5]])
+
+    loss = strategy.compute_loss(relation, child_pos, child_obj.bounding_box, parent_obj.bounding_box)
+    assert loss.shape == (2,)
+    assert loss[0] > 0.0
+    assert torch.isclose(loss[1], torch.tensor(0.0), atol=1e-6)
+
+
+# =============================================================================
+# RelationSolver integration (NotNextTo) tests
```
[the remark this answers is not in the document]
Comment (xyao-nv) [reply isaaclab_arena/tests/test_relation_loss_strategies.py:394]: too many tests, can add later if needed
